rasterio/_features.pyx

```diff
@@ -554,40 +554,52 @@ cdef class OGRGeomBuilder:
             OGR_G_AddGeometryDirectly(geom, part)
         return geom
 
-    cdef OGRGeometryH _buildGeometryCollection(
-            self, object coordinates) except NULL:
+    cdef OGRGeometryH _buildGeomCollection(self, object geoms) except NULL:
         cdef OGRGeometryH part = NULL
-        cdef OGRGeometryH geom = self._createOgrGeometry(
+        cdef OGRGeometryH ogr_geom = self._createOgrGeometry(
             GEOJSON2OGR_GEOMETRY_TYPES['GeometryCollection'])
-        for colxn in coordinates:
-            part = OGRGeomBuilder().build(colxn)
-            OGR_G_AddGeometryDirectly(geom, part)
-        return geom
+        for g in geoms:
+            part = OGRGeomBuilder().build(g)
+            OGR_G_AddGeometryDirectly(ogr_geom, part)
+        return ogr_geom
 
     cdef OGRGeometryH build(self, object geometry) except NULL:
-        """Builds an OGR geometry from GeoJSON geometry."""
+        """Builds an OGR geometry from GeoJSON geometry.
+        Assumes that geometry has been validated prior to calling this; this
+        only does basic checks for validity.
+        """
         cdef object typename = geometry['type']
-        cdef object coordinates = geometry.get('coordinates')
-
-        if not typename or not coordinates:
-            raise ValueError("Input is not a valid geometry object")
-        if typename == 'Point':
-            return self._buildPoint(coordinates)
-        elif typename == 'LineString':
-            return self._buildLineString(coordinates)
-        elif typename == 'LinearRing':
-            return self._buildLinearRing(coordinates)
-        elif typename == 'Polygon':
-            return self._buildPolygon(coordinates)
-        elif typename == 'MultiPoint':
-            return self._buildMultiPoint(coordinates)
-        elif typename == 'MultiLineString':
-            return self._buildMultiLineString(coordinates)
-        elif typename == 'MultiPolygon':
-            return self._buildMultiPolygon(coordinates)
+        cdef object coordinates
```

Comment:
It appeared that assigning this to a member of the dictionary that isn't present in a `GeometryCollection`, and then attempting to later reassign it (original line 589), was not working and was leading to the segfault.